Doc: Porting guide review (#4575)
aristanetworks/ansible-avd: ansible_collections/arista/avd/docs/porting-guides/5.x.x.md

ERROR! [host1]: 'Validation Error: access_lists': {'ACL-01': {'sequence_numbers': {10: {'actio<...abbreviated...>4 any'}}}} is not of type 'list'
```

> As an example, `ethernet_interfaces` would be updated like the following:

Reviewer: ```suggestion
As an example, `ethernet_interfaces` would be updated as follows:
```

```diff
ethernet_interfaces:
-  Ethernet1:   # <-- User defined key
+  - name: Ethernet1 # <-- "name" here is called the "primary_key" which must have a unique value across all list elements
    <...>
```

## Removal of Ansible tags from AVD roles

Support for Ansible tags across AVD roles has been replaced with variables to control the execution of various parts of AVD.

The previous implementation of Ansible tags needed to be more consistent, and they were mostly undocumented. It often led to inconsistent behaviors depending on
when the tags were used with `--tags` vs. `--skip-tags`.

> AVD now supports several variables that allow detailed control of which areas of AVD should be executed.

Reviewer: ```suggestion
AVD now supports several variables that provide detailed control over which areas of AVD should be executed.
```

    | Action plugin/Module | `arista.avd.yaml_templates_to_facts` | The internal `arista.avd.yaml_templates_to_facts` action plugin is no longer used by AVD. The plugin is released as open source, so it can be copied and reused according to the license and copyright. |
    | Action plugin/Module | `arista.avd._emit_warning` | The internal `arista.avd._emit_warning` action plugin is no longer used by AVD. The plugin is released as open source, so it can be copied and reused according to the license and copyright. |

## Changes to role arista.avd.eos_designs

TODO: Level 3 sections for each change with details on how to migrate

### Removed `redistribute igmp` from BGP VLAN config belonging to a VRF with EVPN Multicast

AVD versions below 5.0.0 configure `redistribute igmp` under BGP for all SVIs when `evpn_l2_multicast` is `true`, but configuring `redistribute igmp` with both L2 and L3 EVPN Multicast enabled will take up additional control-plane and data-plane resources.

> In AVD version 5.0.0, by default, `redistribute igmp` is only configured when `evpn_l2_multicast` is `true` and `evpn_l3_multicast` for the VRF is `false`.

Reviewer: ```suggestion
In AVD version 5.0.0, the `redistribute igmp` is configured by default only when `evpn_l2_multicast` is set to `true` and `evpn_l3_multicast` for the VRF is set to `false`.
```

> To retain the previous behavior for SVIs under BGP where both `evpn_l2_multicast` and `evpn_l3_multicast` are enabled set `always_redistribute_igmp` key to `true` either for all SVIs within the Tenant or for each SVIs individually.

Reviewer: ```suggestion
To retain the previous behavior for SVIs under BGP, where both `evpn_l2_multicast` and `evpn_l3_multicast` are enabled, set the `always_redistribute_igmp` key to `true` for either all SVIs within the Tenant or for each SVI individually.
```

    <...>
```

> New (5.0):

Reviewer 2: this is weird - line 41 states `it was changed from`... so `to` makes more sense no?

Reviewer 3: I think this was made before we did the diff-style. Would it make sense to change it completely?

Author: Yeah my bad.  thinking about it "to" signals it existed before it was just updated vs a completely new thing. I really like the diff option. I can make the update

| `arista.avd.eos_config_deploy_cvp` | `avd_eos_config_deploy_cvp_execute_device_tasks` or `execute_tasks` | `false` | Execute pending tasks on CloudVision after onboarding, moving or changing devices. |
| `arista.avd.eos_validate_state` | - | - | See [here](#changes-to-role-aristaavdeos_validate_state) for details on changes in `eos_validate_state`. |

The tasks or functions mentioned will execute when the role is imported/included if the variable is set to `true`. Otherwise, it will not.

This change may require updates to playbooks, "Makefiles" or scripts, to execute the same parts of AVD. It is still possible to skip importing a role by using
Ansible tags in the playbook. The tags will be applied to match all tasks in the imported/included AVD role.

The default behavior is the same as for previous versions of AVD, except for the CSV files, which are no longer generated by default.

For example, to run AVD without generating documentation, the playbook can be updated like this:

```yaml
 - name: Build Configurations and Documentation
   hosts: FABRIC
   gather_facts: false
   tasks:

>      - name: Generate AVD Structured Configurations and Fabric Documentation

Reviewer 2: well techincally 
```suggestion
     - name: Generate AVD Structured Configurations without Fabric Documentation
```

Author: great catch

       ansible.builtin.import_role:
         name: arista.avd.eos_designs
       vars:
         eos_designs_documentation:
           enable: false

>      - name: Generate Device Configurations and Documentation

Reviewer 2: ```suggestion
     - name: Generate Device Configurations without Documentation
```

TODO: Level 3 sections for each change with details on how to migrate

> ### Removed `redistribute igmp` from BGP VLAN configuration belonging to a VRF with EVPN Multicast

Reviewer 2: Changing titles will break release-notes links - please make sure to with them as well (vlan -> VLAN is good but config -> configuration is not)

        <...>
```

> ### The base class for custom IP addressing `AvdIpAddressing` was moved

Reviewer 2: same for titles and release notes link (mkdoc should tell you)

Author: Hmm good point

#### Underlay L2 ethernet interfaces

> The interface description for the L2 uplink/downlink interfaceshas changed to `L2_<peer>_<peer_interface>`. Previously the description was `<peer | upper>_<peer_interface>`.

Reviewer 4: ```suggestion
The interface description for the L2 uplink/downlink interfaces has changed to `L2_<peer>_<peer_interface>`. Previously the description was `<peer | upper>_<peer_interface>`.
```

#### MLAG peer-link port-channel interfaces

The interface description for the MLAG peer-link port-channel interfaces has changed to `MLAG_{mlag_peer}_{peer_interface}`.

> Previously the description was `MLAG_PEER_{mlag_peer}_Po{mlag_port_channel_id}` so using the local port-channel ID instead of the peer's. The MLAG interfaces are often the same, so this is a minor issue.

Reviewer: ```suggestion
Previously, the description was `MLAG_PEER_{mlag_peer}_Po{mlag_port_channel_id}`, so using the local port-channel ID instead of the peer's. The MLAG interfaces are often the same, so this is a minor issue.
```

#### MLAG L3 peering VLAN

> The name of the MLAG L3 peering VLAN (VLAN 4093 by default) has changed to `MLAG_L3`. Previously the name was `LEAF_PEER_L3`.

Reviewer: ```suggestion
The name of the MLAG L3 peering VLAN (VLAN 4093 by default) has changed to `MLAG_L3`. Previously, the name was `LEAF_PEER_L3`.
```

#### MLAG L3 VRF peering VLAN

> The name of the MLAG L3 VRF peering VLAN has changed to `MLAG_L3_VRF_{vrf}`. Previously the name was `MLAG_iBGP_{vrf}`.

Reviewer: ```suggestion
The name of the MLAG L3 VRF peering VLAN has changed to `MLAG_L3_VRF_{vrf}`. Previously, the name was `MLAG_iBGP_{vrf}`.
```

#### MLAG L3 peering

> The description of the MLAG L3 BGP peers, including those in VRFs, has changed to `{mlag_peer}_{peer_interface}`. Previously the description was `{mlag_peer}`.

Reviewer: ```suggestion
The description of the MLAG L3 BGP peers, including those in VRFs, has changed to `{mlag_peer}_{peer_interface}`. Previously, the description was `{mlag_peer}`.
```

#### mpls_route_servers no longer accepts a string

Starting AVD 5.0.0, `<node_type_key>.defaults.mpls_route_servers` only accepts a list of strings.

> Previously, it also accepted a string, which was autoconverted to a list.

Reviewer: ```suggestion
Previously, it also accepted a string, which was auto-converted to a list.
```

Reviewer 5: resolved

      version: 3
```

### Default MTU value changed from `9214` to `9194` for Dps1, WAN HA, and WAN p2p uplink interfaces

In AVD 4.0.0, the default MTU for Dps1, WAN HA, and WAN p2p uplink interfaces was `9214`.

> AVD 5.0.0, changed the default MTU for Dps1, WAN HA, and WAN p2p uplink interfaces to `9194`.

Reviewer: ```suggestion
AVD 5.0.0 has changed the default MTU for Dps1, WAN HA, and WAN p2p uplink interfaces to `9194`.
```

The change has been incorporated into `traffic-policies.j2` eos template.

In AVD 4.0.0, configuring the `flags` for the `UDP` protocol under traffic policies was allowed, but the configuration for `source/destination port` and `source/destination field-set` together was not generated for the `TCP/UDP` protocols.

> As of AVD 5.0.0, the configuration of `flags` in the `UDP` protocol under traffic policies is not generated, and the configuration for `source/destination port` and `field set` can be generated together for the `TCP/UDP` protocol. Also, the configuration order has been fixed as per the EOS running-config.

Author: @carlbuchmann a bit confused here. Should the wording be "is now generated"?

Reviewer 5: good catch, resolved!

The change has been incorporated into `traffic-policies.j2` eos template.

In AVD 4.0.0, configuring the `flags` for the `UDP` protocol under traffic policies was allowed, but the configuration for `source/destination port` and `source/destination field-set` together was not generated for the `TCP/UDP` protocols.

> As of AVD 5.0.0, the configuration of `flags` in the `UDP` protocol under traffic policies is not generated, and the configuration for `source/destination port` and `field set` can be generated together for the `TCP/UDP` protocol. Also, the configuration order has been fixed as per the EOS running-config.

Reviewer 5: ```suggestion
As of AVD 5.0.0, the configuration of `flags` in the `UDP` protocol under traffic policies is now generated, and the configuration for `source/destination port` and `field set` can be generated together for the `TCP/UDP` protocol. Also, the configuration order has been fixed as per the EOS running-config.
```

Reviewer: It should be `not` only, earlier (AVD 4.x)  the `protocol udp flags established` command was generated which is invalid in EOS CLI (`flags` is not supported with `udp`). It has been fixed in this PR - https://github.com/aristanetworks/avd/pull/4266/

Reviewer 5: thanks @Shivani-gslab I re-worded [46002ab](https://github.com/aristanetworks/avd/pull/4575/commits/46002ab6b043e6314d96e4b684fd7fd65ad129fe)
